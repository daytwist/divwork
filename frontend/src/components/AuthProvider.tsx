import Cookies from "js-cookie";
import {
  createContext,
  Dispatch,
  FC,
  ReactNode,
  SetStateAction,
  useContext,
  useMemo,
  useState,
} from "react";

type Props = {
  children: ReactNode;
};

type AuthContextValue = {
  isSignedIn: boolean;
  setIsSignedIn: Dispatch<SetStateAction<boolean>>;
};

export const AuthContext = createContext<AuthContextValue>(
  {} as AuthContextValue
);

export const AuthProvider: FC<Props> = ({ children }) => {
  const [isSignedIn, setIsSignedIn] = useState<boolean>(false);

  const AuthProviderValue = useMemo(
    () => ({ isSignedIn, setIsSignedIn }),
    [isSignedIn, setIsSignedIn]
  );

  return (
    <AuthContext.Provider value={AuthProviderValue}>
      {children}
    </AuthContext.Provider>
  );
};

export const GetIsSignedIn = () => {
  const { isSignedIn, setIsSignedIn } = useContext(AuthContext);

  if (
    Cookies.get("_access_token") &&
    Cookies.get("_client") &&
    Cookies.get("_uid")
  ) {
    setIsSignedIn(true);
    console.log(isSignedIn);
  } else {
    setIsSignedIn(false);
    console.log(isSignedIn);
  }
};
