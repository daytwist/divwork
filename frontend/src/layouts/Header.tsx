const Header: React.FC = () => {

  const onClickSignOut = () => {

  };

  return (
    <div>
      <button type="submit" onClick={onClickSignOut}>
        サインアウト
      </button>
    </div>
  );
};

export default Header;
