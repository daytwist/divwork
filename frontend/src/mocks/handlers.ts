import { rest } from "msw";
import { mockSignIn, mockAuthSessions } from "./mockAuth";
import { mockTeam } from "./mockTeam";
import { mockTeams } from "./mockTeams";
import { mockUser, mockUserEdit } from "./mockUser";

export const handlers = [
  rest.get("/teams/select", mockTeams),
  rest.post("/auth/sign_in", mockSignIn),
  rest.get("/auth/sessions", mockAuthSessions),
  rest.get(`/teams/:id`, mockTeam),
  rest.get(`/users/:id`, mockUser),
  rest.patch(`/auth`, mockUserEdit),
];
