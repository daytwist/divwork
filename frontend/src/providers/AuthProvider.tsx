import { AxiosRequestConfig, AxiosResponse } from "axios";
import Cookies from "js-cookie";
import {
  createContext,
  Dispatch,
  FC,
  ReactNode,
  SetStateAction,
  useEffect,
  useMemo,
  useState,
} from "react";
import { AuthSessionsResponse, User } from "../types/userTypes";
import { axiosInstance } from "../utils/axios";

type Props = {
  children: ReactNode;
};

type AuthContextValue = {
  loading: boolean;
  setLoading: Dispatch<SetStateAction<boolean>>;
  isSignedIn: boolean;
  setIsSignedIn: Dispatch<SetStateAction<boolean>>;
  currentUser: User | undefined;
  setCurrentUser: Dispatch<SetStateAction<User | undefined>>;
  teamReloadFlag: boolean;
  setTeamReloadFlag: Dispatch<SetStateAction<boolean>>;
};

export const AuthContext = createContext<AuthContextValue>(
  {} as AuthContextValue
);

export const AuthProvider: FC<Props> = ({ children }) => {
  const [loading, setLoading] = useState(true);
  const [isSignedIn, setIsSignedIn] = useState(false);
  const [currentUser, setCurrentUser] = useState<User | undefined>(undefined);
  const [teamReloadFlag, setTeamReloadFlag] = useState(false);

  const AuthProviderValue = useMemo(
    () => ({
      loading,
      setLoading,
      isSignedIn,
      setIsSignedIn,
      currentUser,
      setCurrentUser,
      teamReloadFlag,
      setTeamReloadFlag,
    }),
    [
      loading,
      setLoading,
      isSignedIn,
      setIsSignedIn,
      currentUser,
      setCurrentUser,
      teamReloadFlag,
      setTeamReloadFlag,
    ]
  );

  const options: AxiosRequestConfig = {
    url: "/auth/sessions",
    method: "GET",
    headers: {
      "access-token": Cookies.get("_access_token") || "",
      client: Cookies.get("_client") || "",
      uid: Cookies.get("_uid") || "",
    },
  };

  useEffect(() => {
    axiosInstance(options)
      .then((res: AxiosResponse<AuthSessionsResponse>) => {
        console.log(res);

        if (res.data.is_signed_in === true) {
          setIsSignedIn(true);
          setCurrentUser(res.data.current_user);
          setLoading(false);
        } else if (res.data.is_signed_in === false) {
          setIsSignedIn(false);
          setCurrentUser(undefined);
          setLoading(false);
        } else {
          setLoading(true);
        }
      })
      .catch((err) => console.log(err));
  }, [isSignedIn, setCurrentUser]);

  return (
    <AuthContext.Provider value={AuthProviderValue}>
      {children}
    </AuthContext.Provider>
  );
};
