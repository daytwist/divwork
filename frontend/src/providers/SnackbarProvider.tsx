import {
  createContext,
  Dispatch,
  FC,
  ReactNode,
  SetStateAction,
  useMemo,
  useState,
} from "react";
import { AlertColor } from "@mui/material";

type Props = {
  children: ReactNode;
};

type State = {
  open: boolean;
  type: AlertColor | undefined;
  message: string;
};

type SnackbarState = {
  snackbarState: State;
  setSnackbarState: Dispatch<SetStateAction<State>>;
};

export const SnackbarContext = createContext<SnackbarState>(
  {} as SnackbarState
);

export const SnackbarProvider: FC<Props> = ({ children }) => {
  const [snackbarState, setSnackbarState] = useState<State>({
    open: false,
    type: "success",
    message: "",
  });

  const SnackbarStateValue = useMemo(
    () => ({ snackbarState, setSnackbarState }),
    [snackbarState, setSnackbarState]
  );

  return (
    <SnackbarContext.Provider value={SnackbarStateValue}>
      {children}
    </SnackbarContext.Provider>
  );
};
