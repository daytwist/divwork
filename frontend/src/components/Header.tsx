import { AxiosRequestConfig, AxiosResponse } from "axios";
import Cookies from "js-cookie";
import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { axiosInstance } from "../utils/axios";
import { AuthContext } from "./AuthProvider";

const Header: React.FC = () => {
  const { isSignedIn, setIsSignedIn, currentUser } = useContext(AuthContext);

  const navigate = useNavigate();

  const onClickSignOut = () => {
    const options: AxiosRequestConfig = {
      url: "/auth/sign_out",
      method: "DELETE",
      headers: {
        "access-token": Cookies.get("_access_token") || "",
        client: Cookies.get("_client") || "",
        uid: Cookies.get("_uid") || "",
      },
    };

    axiosInstance(options)
      .then((res: AxiosResponse) => {
        console.log(res);

        if (res.status === 200) {
          Cookies.remove("_access_token");
          Cookies.remove("_client");
          Cookies.remove("_uid");

          setIsSignedIn(false);
          navigate("/", { replace: true });
        }
      })
      .catch((err) => console.log(err));
  };

  if (isSignedIn === true) {
    return (
      <div>
        <h4>{currentUser?.name}</h4>
        <button type="submit" onClick={onClickSignOut}>
          サインアウト
        </button>
      </div>
    );
  }

  return (
    <div>
      <h4>サインイン</h4>
    </div>
  );
};

export default Header;
