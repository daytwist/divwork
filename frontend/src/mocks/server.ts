// eslint-disable-next-line import/no-extraneous-dependencies
import { setupWorker } from "msw";
import { handlers } from "./handlers";

export const server = setupWorker(...handlers);
