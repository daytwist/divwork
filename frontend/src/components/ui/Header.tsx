import { useContext } from "react";
import { Link } from "react-router-dom";
import {
  AppBar,
  Toolbar,
  IconButton,
  Typography,
  Box,
  Button,
} from "@mui/material";
import MenuIcon from "@mui/icons-material/Menu";
import { HeaderMenuButton } from "./HeaderMenuButton";
import { AuthContext } from "../../providers/AuthProvider";

type Props = {
  onClick: () => void;
};

export const Header = (props: Props) => {
  const { onClick } = props;
  const { isSignedIn } = useContext(AuthContext);

  return (
    <AppBar
      component="header"
      position="fixed"
      sx={{
        zIndex: {
          xs: 1199,
          lg: 1201,
        },
      }}
    >
      <Toolbar>
        <IconButton
          color="inherit"
          edge="start"
          onClick={onClick}
          sx={{
            flexGrow: 0,
            mr: 2,
            display: { xs: isSignedIn ? "flex" : "none", lg: "none" },
          }}
        >
          <MenuIcon />
        </IconButton>
        <Typography
          variant="h5"
          component={Link}
          to={isSignedIn ? "/teams" : "/"}
          sx={{
            flexGrow: 1,
            display: "block",
            color: "inherit",
            textDecoration: "none",
          }}
        >
          DivWork
        </Typography>
        <div>
          {isSignedIn ? (
            <Box sx={{ flexGrow: 0 }}>
              <HeaderMenuButton />
            </Box>
          ) : (
            <Box sx={{ flexGrow: 0 }}>
              <Button
                component={Link}
                to="/sign_up/teams/select"
                sx={{ my: { xs: 1, md: 2 }, color: "black" }}
              >
                ユーザー登録
              </Button>
              <Button
                component={Link}
                to="/sign_in"
                sx={{ my: { xs: 1, md: 2 }, color: "black" }}
              >
                ログイン
              </Button>
            </Box>
          )}
        </div>
      </Toolbar>
    </AppBar>
  );
};
