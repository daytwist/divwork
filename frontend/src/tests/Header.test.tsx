/* eslint-disable @typescript-eslint/await-thenable */
/* eslint-disable testing-library/no-unnecessary-act */
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { rest } from "msw";
import { act } from "react-dom/test-utils";
import { BrowserRouter } from "react-router-dom";
import Header from "../components/ui/Header";
import { server } from "../mocks/server";
import { AuthProvider } from "../providers/AuthProvider";
import { SnackbarProvider } from "../providers/SnackbarProvider";

describe("Header", () => {
  test("ヘッダーメニュー", async () => {
    server.use(
      rest.get("/auth/sessions", (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            is_signed_in: true,
            current_user: {
              email: "test@example.com",
              uid: "test@example.com",
              id: 1,
              provider: "email",
              allow_password_change: false,
              name: "USER_1",
              nickname: null,
              image: null,
              team_id: 1,
              avatar: "",
            },
          })
        );
      })
    );

    render(
      <BrowserRouter>
        <AuthProvider>
          <SnackbarProvider>
            <Header />
          </SnackbarProvider>
        </AuthProvider>
      </BrowserRouter>
    );

    expect(await screen.findByText("USER_1")).toBeInTheDocument();

    // ユーザー名をクリックするとメニューが表示される
    expect(screen.queryByText("マイタスク")).not.toBeInTheDocument();
    await act(() => {
      userEvent.click(screen.getByText("USER_1"));
    });
    expect(screen.getByText("マイタスク")).toBeInTheDocument();
  });
});
