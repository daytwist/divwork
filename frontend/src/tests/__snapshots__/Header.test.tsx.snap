// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header スナップショット 1`] = `
<header
  className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
>
  <div
    className="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular css-r6ewbb-MuiToolbar-root"
    >
      <a
        className="MuiTypography-root MuiTypography-h5 css-1tn41p-MuiTypography-root"
        href="/"
      >
        DivWork
      </a>
      <div>
        <div
          className="MuiBox-root css-7ln23k"
        >
          <a
            href="/sign_up/teams/select"
            onClick={[Function]}
            style={
              Object {
                "textDecoration": "none",
              }
            }
          >
            <button
              className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1q39md6-MuiButtonBase-root-MuiButton-root"
              disabled={false}
              onBlur={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              サインアップ
            </button>
          </a>
          <a
            href="/sign_in"
            onClick={[Function]}
            style={
              Object {
                "textDecoration": "none",
              }
            }
          >
            <button
              className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1q39md6-MuiButtonBase-root-MuiButton-root"
              disabled={false}
              onBlur={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              サインイン
            </button>
          </a>
        </div>
      </div>
    </div>
  </div>
</header>
`;
